app/src/main/java/ch/epfl/sdp/appart/MapActivity.java

```diff
@@ -115,44 +128,71 @@ private void setupMap() {
-                                        String locality =
-                                                ad.getCity().replaceAll("\\d+", "");
-                                        place = AddressFactory.makeAddress(ad.getStreet(), postalCode, locality);
-
-                                    } else {
-                                        place = LocalityFactory.makeLocality(city);
-                                    }
-                                    return geocodingService.getLocation(place);
-                                }
-                        )
-                                .thenAcceptAsync(location ->
-                                                mapService.addMarker(location
-                                                        , card,
-                                                        false, card.getCity()),
-                                        ContextCompat.getMainExecutor(this));
-                    }
-                });
+                CompletableFuture<Location> futureCurrentLocation =
+                        locationService.getCurrentLocation();
+                CompletableFuture<Stream<Pair<Card, Location>>> filteredAds =
+                        CompletableFuture.allOf(futureCards,
+                                futureCurrentLocation).thenCompose(arg -> {
+                            Location currentLocation =
+                                    futureCurrentLocation.join();
+                            List<Card> cards = futureCards.join();
+                            mapService.zoomOnPosition(currentLocation, 12.0f);
+                            List<CompletableFuture<Pair<Card, Location>>> adsAndLocs =
+                                    cards.parallelStream().map(card -> databaseService.getAd(card.getAdId()).thenCompose(ad -> {
+                                        String city = ad.getCity();
+                                        Matcher extractPostalCodeMatcher =
+                                                Pattern.compile("\\d+").matcher(city);
+                                        Place place;
+                                        try {
+                                            if (extractPostalCodeMatcher.find()) {
+                                                String postalCode =
+                                                        extractPostalCodeMatcher.group();
+                                                String locality =
+                                                        ad.getCity().replaceAll(
+                                                                "\\d" +
+                                                                        "+",
+                                                                "");
+                                                place = AddressFactory.makeAddress(ad.getStreet()
+                                                        , postalCode, locality);
```

Comment:
I'm sorry you had to go through this because of our poor design choice for the `Ad` class